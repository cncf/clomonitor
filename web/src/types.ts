export interface Organization {
  name: string;
  display_name?: string;
  description?: string;
  home_url: string;
  logo_url?: string;
}

export interface BaseProject {
  id: string;
  name: string;
  display_name?: string;
  description?: string;
  home_url: string;
  logo_url?: string;
  devstats_url?: string;
  maturity_id: Maturity;
  category_id: Category;
  score: Score;
  updated_at: number;
}

export interface Project extends BaseProject {
  repositories: BaseRepository[];
  organization: {
    name: string;
  };
}

export interface ProjectDetail extends BaseProject {
  repositories: Repository[];
}

export interface BaseRepository {
  name: string;
  url: string;
  kind: RepositoryKind;
}

export interface Repository extends BaseRepository {
  digest: string;
  repository_id: string;
  score: Score;
  reports: Report[];
}

export interface Report {
  data: CoreReport | any;
  linter_id: LinterId;
  report_id: string;
  updated_at: number;
}

export interface CoreReport {
  // ScoreType
  [key: string]: {
    [key: string]: ReportCheck;
  };
}

export interface ReportCheck {
  passed: boolean;
  value?: string;
  url?: string;
}

export type Score = {
  [key in ScoreType]: number;
} & { score_kind: ScoreKind };

export interface FiltersSection {
  name: string;
  title: string;
  filters: Filter[];
}

export interface Filter {
  name: string | number;
  label: string;
  legend?: string;
  decorator?: JSX.Element;
}

export interface Issue {
  level: number;
  description: string;
}

export interface Prefs {
  search: { limit: number; sort: { by: SortBy; direction: SortDirection } };
  theme: {
    effective: string;
  };
}

export interface ReportOptionData {
  icon: JSX.Element;
  name: string;
  legend: JSX.Element;
}

export enum Maturity {
  Graduated = 0,
  Incubating,
  Sandbox,
}

export enum Category {
  'App definition' = 0,
  Observability,
  Orchestration,
  Platform,
  Provisioning,
  Runtime,
  Serverless,
}

export enum Rating {
  A = 'a',
  B = 'b',
  C = 'c',
  D = 'd',
}

export enum FilterKind {
  Maturity = 'maturity',
  Category = 'category',
  Rating = 'rating',
}

export enum LinterId {
  core = 0,
}

export enum ScoreType {
  BestPractices = 'best_practices',
  Documentation = 'documentation',
  Global = 'global',
  Legal = 'legal',
  License = 'license',
  Security = 'security',
}

export enum SortDirection {
  ASC = 'asc',
  DESC = 'desc',
}

export enum SortBy {
  Name = 'name',
  Score = 'score',
}

export enum RepositoryKind {
  Primary = 'primary',
  Secondary = 'secondary',
}

export enum ScoreKind {
  Primary = 'primary',
  Secondary = 'secondary',
}

export enum ReportOption {
  Adopters = 'adopters',
  ApprovedLicense = 'approved',
  ArtifactHubBadge = 'artifacthub_badge',
  Changelog = 'changelog',
  CodeOfConduct = 'code_of_conduct',
  CommunityMeeting = 'community_meeting',
  Contributing = 'contributing',
  DCO = 'dco',
  Governance = 'governance',
  LicenseScanning = 'scanning',
  Maintainers = 'maintainers',
  OpenSSFBadge = 'openssf_badge',
  Readme = 'readme',
  RecentRelease = 'recent_release',
  Roadmap = 'roadmap',
  SecurityPolicy = 'security_policy',
  SlackPresence = 'slack_presence',
  SPDX = 'spdx_id',
  TrademarkDisclaimer = 'trademark_disclaimer',
  Website = 'website',
}

export interface SearchFiltersURL extends BasicQuery {
  pageNumber: number;
}

export interface BasicQuery {
  text?: string;
  filters?: {
    [key: string]: (string | number)[];
  };
}

export interface SearchQuery extends BasicQuery {
  limit: number;
  offset: number;
  sort_by: SortBy;
  sort_direction: SortDirection;
}

export interface SearchData {
  limit: number;
  offset: number;
  sort_by: string;
  sort_direction: string;
  text?: string;
  category?: number[];
  maturity?: number[];
  rating?: number[];
}

export interface Error {
  kind: ErrorKind;
  message?: string;
}

export enum ErrorKind {
  Other,
  NotFound,
}

export type ReportOptionInfo = {
  [key in ReportOption]: ReportOptionData;
};

export interface RecommendedTemplate {
  name: string;
  url: string;
}
