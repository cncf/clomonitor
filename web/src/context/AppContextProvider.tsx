import { isNull } from 'lodash';
import { createContext, Dispatch, useContext, useEffect, useReducer, useState } from 'react';

import { Prefs, SortBy, SortDirection } from '../types';
import getMetaTag from '../utils/getMetaTag';
import lsStorage from '../utils/localStoragePreferences';
import themeBuilder from '../utils/themeBuilder';

interface AppState {
  prefs: Prefs;
}

interface Props {
  children: JSX.Element;
}

const initialState: AppState = {
  prefs: lsStorage.getPrefs(),
};

type Action =
  | { type: 'updateTheme'; theme: string }
  | { type: 'updateLimit'; limit: number }
  | { type: 'updateSort'; by: SortBy; direction: SortDirection };

export const AppContext = createContext<{
  ctx: AppState;
  dispatch: Dispatch<any>;
}>({
  ctx: initialState,
  dispatch: () => null,
});

export function updateTheme(theme: string) {
  return { type: 'updateTheme', theme };
}

export function updateLimit(limit: number) {
  return { type: 'updateLimit', limit };
}

export function updateSort(by: SortBy, direction: SortDirection) {
  return { type: 'updateSort', by, direction };
}

export function updateActiveStyleSheet(current: string) {
  const secondary = getMetaTag('secondaryColor');
  document.getElementsByTagName('html')[0].setAttribute('data-theme', current);
  document
    .querySelector(`meta[name='theme-color']`)!
    .setAttribute('content', current === 'light' ? secondary : '#0f0e11');
}

export function appReducer(state: AppState, action: Action) {
  let prefs;
  switch (action.type) {
    case 'updateTheme':
      prefs = {
        ...state.prefs,
        theme: {
          effective: action.theme,
        },
      };

      lsStorage.setPrefs(prefs);
      updateActiveStyleSheet(action.theme);
      return {
        ...state,
        prefs: prefs,
      };

    case 'updateLimit':
      prefs = {
        ...state.prefs,
        search: {
          ...state.prefs.search,
          limit: action.limit,
        },
      };
      lsStorage.setPrefs(prefs);
      return {
        ...state,
        prefs: prefs,
      };

    case 'updateSort':
      prefs = {
        ...state.prefs,
        search: {
          ...state.prefs.search,
          sort: {
            by: action.by,
            direction: action.direction,
          },
        },
      };
      lsStorage.setPrefs(prefs);
      return {
        ...state,
        prefs: prefs,
      };

    default:
      return { ...state };
  }
}

function AppContextProvider(props: Props) {
  const activeProfilePrefs = lsStorage.getPrefs();
  const [ctx, dispatch] = useReducer(appReducer, {
    prefs: activeProfilePrefs,
  });
  const [activeInitialTheme, setActiveInitialTheme] = useState<string | null>(null);

  useEffect(() => {
    const theme = activeProfilePrefs.theme.effective || 'light';
    themeBuilder.init();
    updateActiveStyleSheet(theme);
    setActiveInitialTheme(theme);
  }, []); /* eslint-disable-line react-hooks/exhaustive-deps */

  if (isNull(activeInitialTheme)) return null;

  return <AppContext.Provider value={{ ctx, dispatch }}>{props.children}</AppContext.Provider>;
}

function useAppContext() {
  return useContext(AppContext);
}

export { AppContextProvider, useAppContext };
