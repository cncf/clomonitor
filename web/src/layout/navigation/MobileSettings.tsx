import classNames from 'classnames';
import { useContext, useRef, useState } from 'react';
import { FaChartPie, FaFileAlt } from 'react-icons/fa';
import { FiMoon, FiSun } from 'react-icons/fi';
import { HiDotsVertical } from 'react-icons/hi';
import { Link } from 'react-router-dom';

import { AppContext, updateTheme } from '../../context/AppContextProvider';
import useOutsideClick from '../../hooks/useOutsideClick';
import ExternalLink from '../common/ExternalLink';
import styles from './MobileSettings.module.css';

const MobileSettings = () => {
  const [visibleDropdown, setVisibleDropdown] = useState(false);
  const ref = useRef(null);
  const { ctx, dispatch } = useContext(AppContext);
  const { effective } = ctx.prefs.theme;
  useOutsideClick([ref], visibleDropdown, () => setVisibleDropdown(false));

  const onHandleChange = (value: string) => {
    dispatch(updateTheme(value));
    closeDropdown();
  };

  const closeDropdown = () => {
    setVisibleDropdown(false);
  };

  return (
    <div ref={ref} className="d-flex d-md-none ms-auto position-relative">
      <button
        className={`btn btn-sm btn-primary text-white rounded-0 lh-1 ms-3 ${styles.btn}`}
        type="button"
        onClick={() => setVisibleDropdown(!visibleDropdown)}
        aria-label="Mobile settings button"
        aria-expanded={visibleDropdown}
      >
        <HiDotsVertical />
      </button>

      <div role="menu" className={classNames('dropdown-menu rounded-0', styles.dropdown, { show: visibleDropdown })}>
        <div className="px-3 py-2 lightText text-secondary text-uppercase fw-bold">Theme</div>
        <div className="dropdown-item">
          <div className="form-check">
            <input
              id="theme-light"
              name="light"
              className={`form-check-input ${styles.input}`}
              type="radio"
              value="light"
              onChange={() => onHandleChange('light')}
              aria-checked={effective === 'light'}
              tabIndex={-1}
              checked={effective === 'light'}
            />
            <label className={`form-check-label fw-bold w-100 ${styles.label}`} htmlFor="theme-light">
              <FiSun className={`mx-1 position-relative ${styles.icon}`} />
              Light
            </label>
          </div>
        </div>

        <div className="dropdown-item">
          <div className="form-check">
            <input
              id="theme-dark"
              name="dark"
              className={`form-check-input ${styles.input}`}
              type="radio"
              value="dark"
              onChange={() => onHandleChange('dark')}
              aria-checked={effective === 'dark'}
              tabIndex={-1}
              checked={effective === 'dark'}
            />
            <label className={`form-check-label fw-bold w-100 ${styles.label}`} htmlFor="theme-dark">
              <FiMoon className={`mx-1 position-relative ${styles.icon}`} />
              Dark
            </label>
          </div>
        </div>

        <hr />

        <div className="dropdown-item mb-2">
          <ExternalLink
            className="text-decoration-none fw-bold d-inline-block w-100"
            href="/docs"
            label="Open documentation"
            target="_self"
          >
            <div className="d-flex flex-row align-items-center py-1">
              <FaFileAlt />
              <div className="ms-2">Documentation</div>
            </div>
          </ExternalLink>
        </div>

        <div className="dropdown-item mb-2">
          <Link className="text-decoration-none fw-bold d-inline-block w-100" to="/stats" onClick={closeDropdown}>
            <div className="d-flex flex-row align-items-center py-1">
              <FaChartPie />
              <div className="ms-2">Statistics</div>
            </div>
          </Link>
        </div>
      </div>
    </div>
  );
};

export default MobileSettings;
