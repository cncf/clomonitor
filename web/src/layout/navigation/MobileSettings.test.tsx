import { fireEvent, render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { BrowserRouter as Router } from 'react-router-dom';

import { AppContext } from '../../context/AppContextProvider';
import { SortBy, SortDirection } from '../../types';
import MobileSettings from './MobileSettings';

const mockCtx = {
  prefs: {
    search: { limit: 20, sort: { by: SortBy.Name, direction: SortDirection.ASC } },
    theme: { effective: 'light' },
  },
};
const mockDispatch = jest.fn();

describe('MobileSettings', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  it('creates snapshot', () => {
    const { asFragment } = render(
      <AppContext.Provider value={{ ctx: mockCtx, dispatch: jest.fn() }}>
        <Router>
          <MobileSettings />
        </Router>
      </AppContext.Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders proper content', () => {
    render(
      <AppContext.Provider value={{ ctx: mockCtx, dispatch: jest.fn() }}>
        <Router>
          <MobileSettings />
        </Router>
      </AppContext.Provider>
    );

    expect(screen.getByRole('button', { name: 'Mobile settings button' })).toBeInTheDocument();
  });

  it('opens dropdown', () => {
    render(
      <AppContext.Provider value={{ ctx: mockCtx, dispatch: jest.fn() }}>
        <Router>
          <MobileSettings />
        </Router>
      </AppContext.Provider>
    );

    const dropdown = screen.getByRole('menu');
    expect(dropdown).toBeInTheDocument();
    expect(dropdown).not.toHaveClass('show');

    const btn = screen.getByRole('button', { name: 'Mobile settings button' });
    userEvent.click(btn);

    expect(dropdown).toHaveClass('show');
    expect(screen.getByRole('radio', { name: 'Light' })).toBeInTheDocument();
    expect(screen.getByRole('radio', { name: 'Light' })).toBeChecked();
    expect(screen.getByRole('radio', { name: 'Dark' })).toBeInTheDocument();
    expect(screen.getByRole('radio', { name: 'Dark' })).not.toBeChecked();
    expect(screen.getByText('Statistics')).toBeInTheDocument();
    expect(screen.getByRole('link')).toBeInTheDocument();
  });

  it('calls updateTheme event', () => {
    render(
      <AppContext.Provider value={{ ctx: mockCtx, dispatch: mockDispatch }}>
        <Router>
          <MobileSettings />
        </Router>
      </AppContext.Provider>
    );

    const btn = screen.getByRole('button', { name: 'Mobile settings button' });
    userEvent.click(btn);

    fireEvent.click(screen.getByRole('radio', { name: 'Dark' }));

    expect(mockDispatch).toHaveBeenCalledTimes(1);
    expect(mockDispatch).toHaveBeenCalledWith({ theme: 'dark', type: 'updateTheme' });
  });

  it('goes to Stats page', () => {
    render(
      <AppContext.Provider value={{ ctx: mockCtx, dispatch: mockDispatch }}>
        <Router>
          <MobileSettings />
        </Router>
      </AppContext.Provider>
    );

    const btn = screen.getByRole('button', { name: 'Mobile settings button' });
    userEvent.click(btn);

    fireEvent.click(screen.getByRole('link'));

    expect(window.location.pathname).toBe('/stats');
  });
});
