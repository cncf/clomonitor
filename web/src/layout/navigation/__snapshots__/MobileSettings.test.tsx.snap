// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MobileSettings creates snapshot 1`] = `
<DocumentFragment>
  <div
    class="d-flex d-md-none ms-auto position-relative"
  >
    <button
      aria-expanded="false"
      aria-label="Mobile settings button"
      class="btn btn-sm btn-link text-white rounded-0 lh-1 fs-5 ms-3 btn"
      type="button"
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 16 16"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M2.5 12a.5.5 0 0 1 .5-.5h10a.5.5 0 0 1 0 1H3a.5.5 0 0 1-.5-.5m0-4a.5.5 0 0 1 .5-.5h10a.5.5 0 0 1 0 1H3a.5.5 0 0 1-.5-.5m0-4a.5.5 0 0 1 .5-.5h10a.5.5 0 0 1 0 1H3a.5.5 0 0 1-.5-.5"
          fill-rule="evenodd"
        />
      </svg>
    </button>
    <div
      class="dropdown-menu rounded-0 dropdown"
      role="menu"
    >
      <div
        class="px-3 py-2 lightText text-secondary text-uppercase fw-bold"
      >
        Theme
      </div>
      <div
        class="dropdown-item"
      >
        <div
          class="form-check"
        >
          <input
            aria-checked="false"
            class="form-check-input ThemeMode-module_input__iMfv3"
            id="theme-mobile-automatic"
            name="automatic"
            tabindex="-1"
            type="radio"
            value="automatic"
          />
          <label
            class="form-check-label w-100"
            for="theme-mobile-automatic"
          >
            <svg
              class="mx-1 position-relative"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M0 3.75C0 2.784.784 2 1.75 2h20.5c.966 0 1.75.784 1.75 1.75v16.5A1.75 1.75 0 0 1 22.25 22H1.75A1.75 1.75 0 0 1 0 20.25ZM22.5 7h-21v13.25c0 .138.112.25.25.25h20.5a.25.25 0 0 0 .25-.25Zm-10-3.5v2h10V3.75a.25.25 0 0 0-.25-.25ZM7 3.5v2h4v-2Zm-5.25 0a.25.25 0 0 0-.25.25V5.5h4v-2Z"
              />
            </svg>
            Automatic
          </label>
        </div>
      </div>
      <div
        class="dropdown-item"
      >
        <div
          class="form-check"
        >
          <input
            aria-checked="true"
            checked=""
            class="form-check-input ThemeMode-module_input__iMfv3"
            id="theme-mobile-light"
            name="light"
            tabindex="-1"
            type="radio"
            value="light"
          />
          <label
            class="form-check-label w-100"
            for="theme-mobile-light"
          >
            <svg
              class="mx-1 position-relative"
              fill="none"
              height="1em"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <circle
                cx="12"
                cy="12"
                r="5"
              />
              <line
                x1="12"
                x2="12"
                y1="1"
                y2="3"
              />
              <line
                x1="12"
                x2="12"
                y1="21"
                y2="23"
              />
              <line
                x1="4.22"
                x2="5.64"
                y1="4.22"
                y2="5.64"
              />
              <line
                x1="18.36"
                x2="19.78"
                y1="18.36"
                y2="19.78"
              />
              <line
                x1="1"
                x2="3"
                y1="12"
                y2="12"
              />
              <line
                x1="21"
                x2="23"
                y1="12"
                y2="12"
              />
              <line
                x1="4.22"
                x2="5.64"
                y1="19.78"
                y2="18.36"
              />
              <line
                x1="18.36"
                x2="19.78"
                y1="5.64"
                y2="4.22"
              />
            </svg>
            Light
          </label>
        </div>
      </div>
      <div
        class="dropdown-item"
      >
        <div
          class="form-check"
        >
          <input
            aria-checked="false"
            class="form-check-input ThemeMode-module_input__iMfv3"
            id="theme-mobile-dark"
            name="dark"
            tabindex="-1"
            type="radio"
            value="dark"
          />
          <label
            class="form-check-label w-100"
            for="theme-mobile-dark"
          >
            <svg
              class="mx-1 position-relative"
              fill="none"
              height="1em"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M21 12.79A9 9 0 1 1 11.21 3 7 7 0 0 0 21 12.79z"
              />
            </svg>
            Dark
          </label>
        </div>
      </div>
      <hr />
      <div
        class="dropdown-item mb-2"
      >
        <a
          aria-label="Open documentation"
          class="link text-decoration-none fw-bold d-inline-block w-100"
          href="/docs"
          rel="noopener noreferrer"
          tabindex="-1"
          target="_self"
        >
          <div
            class="d-flex flex-row align-items-center py-1"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 384 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M224 136V0H24C10.7 0 0 10.7 0 24v464c0 13.3 10.7 24 24 24h336c13.3 0 24-10.7 24-24V160H248c-13.2 0-24-10.8-24-24zm64 236c0 6.6-5.4 12-12 12H108c-6.6 0-12-5.4-12-12v-8c0-6.6 5.4-12 12-12h168c6.6 0 12 5.4 12 12v8zm0-64c0 6.6-5.4 12-12 12H108c-6.6 0-12-5.4-12-12v-8c0-6.6 5.4-12 12-12h168c6.6 0 12 5.4 12 12v8zm0-72v8c0 6.6-5.4 12-12 12H108c-6.6 0-12-5.4-12-12v-8c0-6.6 5.4-12 12-12h168c6.6 0 12 5.4 12 12zm96-114.1v6.1H256V0h6.1c6.4 0 12.5 2.5 17 7l97.9 98c4.5 4.5 7 10.6 7 16.9z"
              />
            </svg>
            <div
              class="ms-2"
            >
              Documentation
            </div>
          </div>
        </a>
      </div>
      <div
        class="dropdown-item mb-2"
      >
        <a
          class="text-decoration-none fw-bold d-inline-block w-100"
          href="/stats"
        >
          <div
            class="d-flex flex-row align-items-center py-1"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 544 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M527.79 288H290.5l158.03 158.03c6.04 6.04 15.98 6.53 22.19.68 38.7-36.46 65.32-85.61 73.13-140.86 1.34-9.46-6.51-17.85-16.06-17.85zm-15.83-64.8C503.72 103.74 408.26 8.28 288.8.04 279.68-.59 272 7.1 272 16.24V240h223.77c9.14 0 16.82-7.68 16.19-16.8zM224 288V50.71c0-9.55-8.39-17.4-17.84-16.06C86.99 51.49-4.1 155.6.14 280.37 4.5 408.51 114.83 513.59 243.03 511.98c50.4-.63 96.97-16.87 135.26-44.03 7.9-5.6 8.42-17.23 1.57-24.08L224 288z"
              />
            </svg>
            <div
              class="ms-2"
            >
              Statistics
            </div>
          </div>
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
